### PR TITLE
[SPARK-7135][SQL] DataFrame expression for monotonically increasing IDs.

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

**Comment by pwendell** [comment]
Could it be confusing to users that the ID associated with each record might be different on stage or task retries? The fact that ordering within a partition is not deterministic has caused people some concern in the past, and I wonder if this could sort of lead to more confusion since you are giving some _sort of_ ordering semantics.

**Comment by rxin** [comment]
partition id doesn't change between retries, does it?

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -103,8 +103,28 @@ def countDistinct(col, *cols):
     return Column(jc)
 
 
+def monotonicallyIncreasingId():
+    """A column that generates monotonically increasing 64-bit integers.
+
+    The generated ID is guaranteed by be monotonically increasing and unique, but not consecutive.
```
Comment (rxin) [inline python/pyspark/sql/functions.py]: guaranteed TO be

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -103,8 +103,28 @@ def countDistinct(col, *cols):
...
+    The current implementation puts the partition ID in the upper 31 bits, and the record number
+    within each partition in the lower 33 bits. The assumption is that the data frame has
+    less than 1 billion partitions, and each partition has less than 20 billion records.
```
Comment (rxin) [inline python/pyspark/sql/functions.py]: 20 b -> 4b ...

**Comment by pwendell** [comment]
No, but the ordering of records in a partition can change, so you might have different identifiers for the same record across retries (unless this is only used for already sorted data... is it?).

**Comment by rxin** [comment]
Those could change in shuffle I guess, but I don't think this is creating more confusion. What we care about here is not the record ordering, but the output of this expression is monotonic increasing. That will always be true.

This is very similar to the row id idea a lot of databases have. SQL tables also don't have ordering, unless they are sorted.

**Comment by rxin** [comment]
@pwendell you raised a very good point about ordering of records within RDDs and DataFrames. I think we should document those more clearly in the javadoc for these.

**Comment by pwendell** [comment]
@rxin yeah I just mean if I'm in a database and I run the same query twice, I will get the same row ID for the same record. Because of non determinism in the shuffle, that's not true here.

**Comment by rxin** [comment]
(That's not always true -- somebody could've deleted an index and then the scan gets turned from index scan to sequential scan, and then record ordering changed)

**Comment by pwendell** [comment]
Oh I see - I guess it doesn't matter then.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by yhuai** [comment]
LGTM

**Pull request closed without merging**